platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/WorkNodeCodec.kt

```diff
@@ -55,81 +76,319 @@ typealias IdForNode = (Node) -> Int
+    ): MutableList<Int> {
         // Not all entry nodes are always scheduled.
         // In particular, it happens when the entry node is a task of the included plugin build that runs as part of building the plugin.
         // Such tasks do not rerun when configuration cache is re-used, even if specified on the command line.
         // Not restoring them as entry points doesn't affect the resulting execution plan.
         val scheduledEntryNodeIds = mutableListOf<Int>()
         nodes.forEach { node ->
-            write(node)
             val nodeId = scheduledNodeIds.size
             scheduledNodeIds[node] = nodeId
-            if (node in work.entryNodes) {
+            if (node in entryNodes) {
                 scheduledEntryNodeIds.add(nodeId)
             }
             if (node is LocalTaskNode) {
                 scheduledNodeIds[node.prepareNode] = scheduledNodeIds.size
             }
         }
-        // A large build may have many nodes but not so many entry nodes.
-        // To save some disk space, we're only saving entry node ids rather than writing "entry/non-entry" boolean for every node.
-        writeCollection(scheduledEntryNodeIds) {
-            writeSmallInt(it)
+        return scheduledEntryNodeIds
+    }
+
+
+    private
+    fun WriteContext.writeNodes(
+        nodes: ImmutableList<Node>,
+        nodeIds: Object2IntOpenHashMap<Node> // Map<Node, NodeId>
+    ): Map<ActionNode, List<Node>> {
+        writeSmallInt(nodeIds.size)
+
+        val groupedNodes = nodes.groupBy(NodeOwner::of)
+        writeCollection(groupedNodes.keys) { nodeOwner ->
+            val groupPath = nodeOwner.path()
+            writeString(groupPath.path)
         }
-        val idForNode: IdForNode = scheduledNodeIds::getInt
-        nodes.forEach { node ->
-            writeSuccessorReferencesOf(node, idForNode)
-            writeNodeGroup(node.group, idForNode)
+
+        val partialResultsRef = AtomicReference<PersistentList<List<PostExecutionNodes>>>(PersistentList.of())
+
+        // TODO:parallel-cc is this message useful for the context where it may show?
+        runBuildOperations(parallelStore, "saving state nodes" ) {
+            groupedNodes.entries.map { (nodeOwner, groupNodes) ->
+                val groupPath = nodeOwner.path()
+                OperationInfo(displayName = "Storing $groupPath", progressDisplayName = groupPath.path) {
+                    contextSource.writeContextFor(this@writeNodes, groupPath).useToRun {
+                        val postExecutionSuccessors = writeGroupedNodes(nodeOwner, groupNodes, nodeIds)
+                        if (postExecutionSuccessors.isNotEmpty()) {
+                            partialResultsRef.updateAndGet {
+                                it.plus(postExecutionSuccessors)
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        val result = IdentityHashMap<ActionNode, List<Node>>()
```

Comment:
💅 This could be a separate function `return aggregatePostExecutionNodes(partialResultsRef.get())`